### ISSUE TITLE
Make pipeline modules function with prefix tags

### ISSUE THREAD

**Issue by ftyers** [apertium/apertium#106]
At the moment some modules in the pipeline require that all tags come *after* the lemma. For some languages it is much more convenient to have the tags before, e.g. for languages that are primarily, or significantly prefixing (e.g. Mayan or Nahuan languages). When translating between these languages, it is a bit of a faff and introduces possibility of errors to have the tags first prefixed, then moved to the end, then have to move them back to the beginning.

```
$ echo "nutinamit" | apertium -d ../apertium-quc/ quc-morph
^nutinamit/<px1sg>tinamit<n>$^./.<sent>$

$ echo "intinmit" | apertium -d ../apertium-usp/ usp-morph
^intinmit/<px1sg>tinmit<n>$^./.<sent>$
```

Going `nutinamit ` → `<px1sg>tinamit<n>` → `tinamit<n><px1sg>` → `tinmit<n><px3sg>` → `<px1sg>tinmit<n>` → `intinmit` seems a bit excessive.

**Comment by mlforcada** [comment]
Completely agree.
lttools finite-state processors can deal with tags anywhere as far as I know.
Which modules in particular create trouble? Transfer, I assume, but not sure.

**Comment by flammie** [comment]
I think I filed a bug before when I was workign with a student with bas-btc, perhaps it was just  VISL CG 3 had issues with prefixed tags... 

I wrote a wiki page on how to deal with prefixes then, maybe this one: https://wiki.apertium.org/wiki/Lexc_and_flag_diacritics_for_prefix_tagging

**Comment by TinoDidriksen** [comment]
CG had issues with prefix tags before 2020-04-01. Since then it moves them to the end. But that is definitely not the best solution, so https://github.com/TinoDidriksen/cg3/issues/64

**Comment by khannatanmai** [comment]
The apertium tagger used to give a parsing error with prefixed tags. In https://github.com/apertium/apertium/commit/2284697929b3d713fc67197cf59b50e95598d565 I modified it to allow prefix tags, but because of the way it was written, it accepts the input, but converts all to suffix tags. Preserving prefix tags and suffix tags will require a deeper change in the code.

**Comment by TinoDidriksen** [comment]
For Greenlandic, we move them to immediately after the baseform, e.g. the output from raw analysis is
````
$ echo 'tamatumuunakkut' | hfst-optimized-lookup -p -u /usr/share/giella/kal/analyser-disamb-gt-desc.hfstol
tamatumuunakkut TA+manna+Gram/Dem+Pron+Via+Sg
````
which when transformed to CG becomes
````
$ echo 'tamatumuunakkut' | kal-tokenise
"<tamatumuunakkut>"
        "manna" Prefix/TA Gram/Dem Pron Via Sg
````

This lets the rest of the pipe know they are prefixes, and they're before all the other morphemes. Only the final generation step moves them back.

**Comment by mr-martian** [comment]
-separable has essentially the same formalism as monodix and includes an explicit way to write `<ANY_TAG>`, so it should work just fine.

-recursive has roughly the same issues as t*x, though once a syntax for expressing them was decided on, I could probably make the relevant changes in no more than an hour or two.

-biltrans sounds like it would benefit some sort of explicit `<ANY_TAG>`, though I haven't thought through what form that might take.

lexical selection as also a relatively straightforward fix.

Regarding the taggers, from digging through them a couple weeks ago, I already had the urge to significantly rewrite them if/when a new file format was decided on.

Regarding flag diacritics, might I recommend lexd? :)

**Comment by flammie** [comment]
> > I think I filed a bug before when I was workign with a student with bas-btc, perhaps it was just VISL CG 3 had issues with prefixed tags...
> > I wrote a wiki page on how to deal with prefixes then, maybe this one: https://wiki.apertium.org/wiki/Lexc_and_flag_diacritics_for_prefix_tagging
> 
> Yep, that's possible, it's also possible to do it as a postprocess using rewrite rules. But both of them kind of suck. If you do it as a postprocess your transducer topology gets all out of whack (but maybe `hfst-realign` can help with that now), if you do it like you suggest, the source code looks like a massive mess D:

I agree, what I described was just a hack around the limitations I came up on the spot, so to say, though lexc altogether is way past expiration date. maybe lexd will be less European-suffixing-lang-centric design replacement for us in the future?

**Comment by unhammer** [comment]
Francis Tyers <notifications@github.com> čálii:

> * Lexical transfer: Works ok, but because of how lookup works,
> it's necessary to include all of the potential prefixes in the
> bilingual dictionary as paradigms. See [1], where the `<par
> n="N"/>` lists all the possible possessive prefixes. I think some
> better solution is possible here, although I don't know what it is
> yet.

Would it be expensive for fst_processor to put epsilon arcs from root to
the start of any lemma? (ie. skipping prefix tags) Also from the points
in between prefix tags, in case one wants to list things with a required
prefix tag.

**Comment by ftyers** [comment]
Not sure, but yeah, I think we'd definitely want to be able to require prefix tags too, although perhaps only right next to the lemma (more or less how the lexical transfer currently works).

**Comment by mr-martian** [comment]
What about postprocessing with a separate transducer?

```xml
<dictionary type="separable">
  <alphabet>blah</alphabet>
  <sdefs><sdef n="n"/><sdef n="px1sg"/></sdefs>
  <section id="main" type="standard">
    <e>
      <p><l><s n="px1sg"/></l><r></r></p>
      <i><w/><t/></i>
      <p><l></l><r><s n="px1sg"/></r></p>
      <i><j/></i>
    </e>
  </section>
</dictionary>
```

```bash
daniel@computer-name-here:~$ lsx-comp lr prefix.lsx prefix.bin
main@standard 8 10
daniel@computer-name-here:~$ echo "^<px1sg>tinamit<n>$" | lsx-proc prefix.bin
^tinamit<n><px1sg>$
daniel@computer-name-here:~$ lsx-comp rl prefix.lsx unprefix.bin
main@standard 8 10
daniel@computer-name-here:~$ echo "^tinamit<n><px1sg>$" | lsx-proc unprefix.bin 
^<px1sg>tinamit<n>$
```

**Comment by mr-martian** [comment]
see https://github.com/apertium/apertium-zul/commit/2a011974f0aa39087ed9d344882fa4fe3a738893 for my rearranging idea in practice.
